### PR TITLE
Issue #120: Add PreferMethodReferenceCheck

### PR DESCRIPTION
**Pull request by strkkk**
#120

Reg report - https://strkkk.github.io/120/
Contrib PR - https://github.com/checkstyle/contribution/pull/384

### CODE REVIEW

**Comment by strkkk** [comment]
@rnveach Can you tell me what is wrong with wercker build?

**Comment by romani** [comment]
From wercker:
>.CheckstyleException: Unable to instantiate 'OrderedProperties' class

Please recheck one more than can not see long from wercker, for me it opens logs anymously

**Comment by strkkk** [comment]
> Please recheck one more than can not see long from wercker, for me it opens logs anymously

it shows 401 for me even if logged in

**Comment by romani** [comment]
> it shows 401 for me even if logged in

yes, just tested it. 
Please register at wercker and send me your nickname at wercker (usually same as gihub, but I donot see you) I will add you "contributors" team to be able to view logs at https://app.wercker.com/organizations/checkstyle/settings/teams/5d1318becde21d1400080fe0

@pbludov , I see your account in wercker and I shared with you permission to logs, please confirm that you have access.

**Comment by strkkk** [comment]
@romani my nickname there is "strkk" (2 k instead of 3 in github)

**Comment by pbludov** [comment]
> @pbludov , I see your account in wercker and I shared with you permission to logs, please confirm that you have access.

Yes, I have. And succeeded to restart the wercker pipeline for this PR. So, the necessary permissions are available to me.

**Comment by romani** [comment]
>my nickname there is "strkk"

Please recheck access

**Comment by strkkk** [comment]
> Please recheck access

It is ok now.

**Comment by strkkk** [comment]
@pbludov can you review this?

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/PreferMethodReferenceCheck.java [hunk 1]
```diff
@@ -0,0 +1,337 @@
...
+
+    /**
+     * A key is pointing to the warning message text in "messages.properties"
+     * file.
+     */
+    public static final String MSG_METHOD_REF = "method.reference";
+
+    /**
+     * List of tokens that is used to check expression.
+     * @see #shouldCheck(DetailAST)
```
Comment (pbludov) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/PreferMethodReferenceCheck.java]: Please add a blank line before the tags section as suggested in #4356
We did not enforce this right now, but in this PR part of comments follow this rule, and others do not. Please make them all consistent.

File: src/main/resources/com/puppycrawl/tools/checkstyle/checks/coding/messages_de.properties [hunk 2]
```diff
@@ -27,6 +27,7 @@ illegal.type=Die Verwendung des Typs ''{0}'' ist nicht erlaubt.
 inline.conditional.avoid=Der Inline-Conditional-Operator sollte vermieden werden.
 instantiation.avoid=Die Klasse {0} sollte nicht direkt instanziiert werden.
 magic.number=Die magische Zahl ''{0}'' sollte als Konstante definiert werden.
+method.reference=Lambda can be replaced with method reference.
```
Comment (pbludov) [inline src/main/resources/com/puppycrawl/tools/checkstyle/checks/coding/messages_de.properties]: Lambda kann durch methodenreferenz ersetzt werden.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/prefermethodreference/InputPreferMethodReferenceMethodCalls.java [hunk 3]
```diff
@@ -0,0 +1,80 @@
+package com.puppycrawl.tools.checkstyle.checks.coding.prefermethodreference;
+
+import static java.lang.Long.sum;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;import java.util.Optional;import java.util.concurrent.Callable;
```
Comment (pbludov) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/prefermethodreference/InputPreferMethodReferenceMethodCalls.java]: Why these imports are messed up? Imports have no effect to this check.
If this is in order not to alter a large number of asserts, you can move the first two cases to the end - the rest will remain in place.

File: src/xdocs/config_coding.xml [hunk 4]
```diff
@@ -2669,7 +2669,7 @@ class TestMethodCall {
       </subsection>
     </section>
 
-    <section name="MissingCtor">
+      <section name="MissingCtor">
```
Comment (pbludov) [inline src/xdocs/config_coding.xml]: Your IDE did something you didn't expect. Please, redo this indentation.

**Comment by pbludov** [comment]
@strkkk, there are no xpath suppression checks in this PR. Are you planning to add them?

File: src/xdocs/config_coding.xml (hunk 4 above)
Comment (strkkk) [reply src/xdocs/config_coding.xml]: corrected.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/prefermethodreference/InputPreferMethodReferenceMethodCalls.java (hunk 3 above)
Comment (strkkk) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/prefermethodreference/InputPreferMethodReferenceMethodCalls.java]: just garbage, deleted.

File: src/main/resources/com/puppycrawl/tools/checkstyle/checks/coding/messages_de.properties (hunk 2 above)
Comment (strkkk) [reply src/main/resources/com/puppycrawl/tools/checkstyle/checks/coding/messages_de.properties]: corrected.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/PreferMethodReferenceCheck.java (hunk 1 above)
Comment (strkkk) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/PreferMethodReferenceCheck.java]: done

**Comment by strkkk** [comment]
> @strkkk, there are no xpath suppression checks in this PR. Are you planning to add them?

Yes, I do.

But I don't know what test cases add to xpath suppression, since there are many of then in original cases.
Can you share your opinion on it?

**Comment by strkkk** [comment]
@pbludov I corrected all things mentioned in your review. For some reason, github doesn't show it and ci wasn't triggered.
Actual version of changes is here https://github.com/strkkk/checkstyle/commit/0e6109c4f736bc1921c8003f9bacc7c3fcbade45

**Comment by rnveach** [comment]
> But I don't know what test cases add to xpath suppression, since there are many of then in original cases.
Can you share your opinion on it?

It doesn't need to be as detailed as original testing. Just ensure every `log` call is validated once.
You can look at other xpath tests for examples.

**Comment by romani** [comment]
Regression report is interesting.

Can you send PR to Vavr project (select several different by style code) with proposal to update code ?
I am not sure they will like such update, so we need to know experts feedback on such Check.

**Comment by strkkk** [comment]
@romani 
I won't have time for all this PRs (spring, vavr and spotbugs). I can create issues there with report link.

I will send pr to sevntu checkstyle later.

**Comment by romani** [comment]
>I won't have time for all this PRs 

By means of such approach, PRs to project to test Check maturity to work on real code. If users reject to apply fixes and explain why, it is food for thoughts for us on how to improve Check.
We should not fix all their problems, but we should try several for sure.

I see this
https://strkkk.github.io/120/Vavr/index.html#A1 as false positive, you think it is valid, please prove that Check is right.
Accepted PRs is the best prove on this :).

**Comment by strkkk** [comment]
> I see this
> https://strkkk.github.io/120/Vavr/index.html#A1 as false positive, you think it is valid, please prove that Check is right.


Original code:
```
.thenComparing(t -> includeNames.indexWhere(i -> t.getImplementation().startsWith(i)));
```

Refactored code:
```
.thenComparing(t -> includeNames.indexWhere(t.getImplementation()::startsWith));
```

Check works.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/prefermethodreference/InputPreferMethodReferenceObjectCreation.java [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+import java.util.function.Supplier;
+
+/**
+ * Config = default
+ * Object/Array creation cases
+ */
+public class InputPreferMethodReferenceObjectCreation {
+    int field;
+    // object creation
+    Supplier<Object> s = () -> new Object(); // violation
```
Comment (romani) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/prefermethodreference/InputPreferMethodReferenceObjectCreation.java]: please add test cases for (with and without curly braces):
```
() -> { return new HashSet<>(); }
() -> { return new HashSet<String>(); }
```
to be sure that Check behavior is not affected by generics.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/PreferMethodReferenceCheck.java [hunk 6]
```diff
@@ -0,0 +1,344 @@
...
+     * @param declarator array declarator to check
+     * @param name expr ident to compare against
+     * @return {@code true} if first array init expression is {@link TokenTypes#IDENT} and
+     *     same as {@code name} and there are no more expressions (if array is multidimensional),
+     * {@code false} otherwise
+     */
+    private static boolean isArgUsedInArrayInit(DetailAST declarator, String name) {
+        final DetailAST nestedDecl = declarator.findFirstToken(TokenTypes.ARRAY_DECLARATOR);
+        final DetailAST expr = declarator.findFirstToken(TokenTypes.EXPR);
+        boolean result = false;
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/PreferMethodReferenceCheck.java]: can it be `final boolean result;` ?

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/PreferMethodReferenceCheck.java [hunk 7]
```diff
@@ -0,0 +1,344 @@
...
+    /**
+     * Check that creation of the object can be replaced with constructor reference.
+     *
+     * @param ast token to check
+     * @param args lambda arguments
+     * @return true if token can be replaced, false otherwise
+     */
+    private static boolean checkObjectCreation(DetailAST ast, List<String> args) {
+        final DetailAST elist = ast.findFirstToken(TokenTypes.ELIST);
+        final DetailAST arrayDecl = ast.findFirstToken(TokenTypes.ARRAY_DECLARATOR);
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/PreferMethodReferenceCheck.java]: can we split this method into two methods?
 it look like check two separate concerns: object and array

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/PreferMethodReferenceCheck.java [hunk 8]
```diff
@@ -0,0 +1,344 @@
...
+            case TokenTypes.LITERAL_NEW:
+                result = checkObjectCreation(child, args);
+                break;
+            case TokenTypes.SLIST:
+            case TokenTypes.LITERAL_RETURN:
+                result = child.getFirstChild().getType() != TokenTypes.SEMI
+                    && checkExpr(child.getFirstChild(), args);
+                break;
+            default:
+                result = shouldCheck(expr)
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/PreferMethodReferenceCheck.java]: please order method declaration by our future Check - https://github.com/sevntu-checkstyle/methods-distance
to minimize distance between methods (for now it is web service).

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/PreferMethodReferenceCheck.java [hunk 9]
```diff
@@ -0,0 +1,344 @@
...
+     * file.
+     */
+    public static final String MSG_METHOD_REF = "method.reference";
+
+    /**
+     * List of tokens that is used to check expression.
+     *
+     * @see #shouldCheck(DetailAST)
+     */
+    private static final List<Integer> TOKEN_TYPES = Arrays.asList(
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/PreferMethodReferenceCheck.java]: can we give less generic name to this collection ?
BLOCK_TOKEN_TYPES ?

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/PreferMethodReferenceCheck.java (hunk 8 above)
Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/PreferMethodReferenceCheck.java]: @rnveach  and @pbludov , Looks like we forgot about our experimental attempt to control even order of methods in class. While we are in progress of searching good formula for order, we try to "enforce" such "Check" over new Modules. Please request this during review of new Modules.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/PreferMethodReferenceCheck.java (hunk 6 above)
Comment (strkkk) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/PreferMethodReferenceCheck.java]: It is false by default.
I can make it final and add else block with `result = false`, but in my opinion it is better to leave it like this.

Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/PreferMethodReferenceCheck.java]: Agree, I was not attentive.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/PreferMethodReferenceCheck.java [hunk 10]
```diff
@@ -0,0 +1,344 @@
...
+import com.puppycrawl.tools.checkstyle.api.DetailAST;
+import com.puppycrawl.tools.checkstyle.api.TokenTypes;
+import com.puppycrawl.tools.checkstyle.utils.TokenUtil;
+
+/**
+ * <p>
+ * Checks that lambda usage can be replaced with method reference.
+ * </p>
+ * <p>
+ * Rationale: method references are easier to read.
```
[the remark this answers is not in the document]
Comment (strkkk) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/PreferMethodReferenceCheck.java]: done

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/prefermethodreference/InputPreferMethodReferenceObjectCreation.java [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+    Function<Integer, Object> oc1 = a -> new int[a]; // violation
+    Function<Integer, Object> oc2 = a -> new int[a][]; // violation
+    Function<String, Example> oc3 = a -> new Example(a); // violation
+    BiFunction<String, String, Example> oc4 = (a, b) -> new Example(a, b); // violation
+
+    // not violations
+    Function<String, Object> noArg = a -> new Object();
```
[the remark this answers is not in the document]
Comment (strkkk) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/prefermethodreference/InputPreferMethodReferenceObjectCreation.java]: This is same thing as comment below 
https://github.com/checkstyle/checkstyle/pull/6846#discussion_r301379863

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/prefermethodreference/InputPreferMethodReferenceObjectCreation.java (hunk 5 above)
Comment (strkkk) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/prefermethodreference/InputPreferMethodReferenceObjectCreation.java]: done

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/PreferMethodReferenceCheck.java (hunk 7 above)
Comment (strkkk) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/PreferMethodReferenceCheck.java]: done

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/PreferMethodReferenceCheck.java (hunk 8 above)
Comment (strkkk) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/PreferMethodReferenceCheck.java]: I rearranged them, please take a look

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/PreferMethodReferenceCheck.java (hunk 9 above)
Comment (strkkk) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/PreferMethodReferenceCheck.java]: changed

**Comment by strkkk** [comment]
> Unfortunately I have to insist on PR to VAVR project (maintainers of such project are over love functional style, we need to know why they missed method ref in code).

Why issue is not enough? We need just a feedback.

> Even technically your Check is right, there are cases where I do not see method reference make code more readable (arguments are words that helps to explicitly define context).

Yes, but I do not see how Check can distinguish readable from not-readable.

**Comment by romani** [comment]
>Why issue is not enough? We need just a feedback.

If it work, I am fine.
But in my experience, PR work much better, feedback will be quicker.

>but I do not see how Check can distinguish readable from not-readable.

Let's see where users reject to follow refactoring. It is not the first time we deal with something like this

**Comment by strkkk** [comment]
I created an issue there - https://github.com/vavr-io/vavr/issues/2445

**Comment by romani** [comment]
Lets wait for 1 week for feedback.
If no feedback happen, I recommend to send PR.

**Comment by romani** [comment]
@KTannenberg, please share your ideas on Vavr diff report (see link in PR description). 
How to you like new Check, do you see any cases, where it is better to keep lambda ? Can you run new Check on your code base and give us feedback.

**Comment by strkkk** [comment]
@romani May be let user decide what parameter is readable and what not be providing regexp for parameter names?
If both params match regex, we can skip this lambda. Otherwise, report a violation.
```
// regexp = [a-z][a-zA-Z]{2,}
Function<Object, Object> f = someName -> someName.toString(); // OK
Function<Object, Object> f1 = o -> o.toString(); // violation
BiFunction<Long, Long, Long> bf = (a, someParam) -> Long.sum(a,someParam); // violation, only one arg matches
BiFunction<Long, Long, Long> bf = (firstParam, someParam) -> Long.sum(firstParam, someParam); // OK, both args match
```

**Comment by strkkk** [comment]
@romani there was some feedback in issue regarding cases when invocation target is expression. There is a difference between lambda and method ref. Such cases can be excluded (by default or permanently) or we can add a note that this can change semantics (IDEA has warning for such cases)

**Comment by romani** [comment]
Please make a summary of false positives and won't fixes to let us think think about what can we do.

**Comment by strkkk** [comment]
> Please make a summary of false positives and won't fixes to let us think think about what can we do.

1. Violations in tests are irrelevant (nothing we can do)
2. Expressions can be replaced, but can change semantics (make it not lazy), e.g.
``` a -> myField.method().otherMethod(a)```
can be replaced with
``` myField.method()::otherMethod ```
but this would cause evaluation of `myField.method()` before lambda invocation, so it won't be lazy. So these changes are rejected.
I dont see big problem in it, it can be done with additional property (detectExpressions, disabled by default) or excluded from check.
3. No comments about readability issues

**Comment by romani** [comment]
>Violations in tests are irrelevant 

Code is relevant, but author has no interest to pay attention to it, sad but reasonable from his perspective.

>Expressions can be replaced, but can change semantics (make it not lazy), 

We need new property on this, no violation on expressions by default, user need to activate it and understand possible side effects.

>No comments about readability issues

Yes, in response, there clear message that, they do not care about this at all.

----

After Check update please provide two reports, to compare behavior with and without the property activation.

**Comment by strkkk** [comment]
> We need new property on this, no violation on expressions by default, user need to activate it and understand possible side effects

What is a name for this property?
Something like `detectExpressions`?

**Comment by romani** [comment]
Name is big struggle.
What about `detectInExpression` or `targetExpression`.
Please choose any name and let's proceed. It will be more clear what name is better on diff report with and without activation of such property.

**Comment by romani** [comment]
please bump version of checkstyle in doc.
Sorry for long acceptance process.

**Comment by romani** [comment]
@strkkk , ping.

**Comment by strkkk** [comment]
I will finish it (version bump as well)
I do not have much time now for it.

**Comment by romani** [comment]
> I do not have much time now for it.

thanks a lot for hint, we appreciate your contributions.

**Comment by Vampire** [comment]
I'd like to suggest for the property in question to not mention expression, but a more generic name like (bad name, just to get the idea) `searchForProbablyChangingSemanticCases`.

The problem is not only with called methods, but also with accessed fields.
If you use local variables in lambdas, they have to be effectively final or compilation will fail.
But for fields this is different, they can freely be modified.

This means the example lambda from above `a -> myField.method().otherMethod(a)` has two problematic things, the field and the method, assuming the field is really a field and not a local variable.

If you use `a -> myField.otherMethod(a)`, then at the time the lambda is **used** the current value of `myField` is used to call `otherMethod`.
If you replace it with `myField::otherMethod`, then at the time the method reference is **evaluated**  the value of `myField` is captured and everytime the lambda is _used_ `otherMethod` is called on that captured instance, no matter what value `myField` has at that time.

If you use `a -> myMethod().otherMethod(a)`, then at the time the lambda is **used** `myMethod` is freshly called and its return value used to call `otherMethod`.
If you replace it with `myMethod()::otherMethod`, then at the time the method reference is **evaluated**  `myMethod` is called and its return value is captured and everytime the lambda is _used_ `otherMethod` is called on that captured instance, no matter what the `myMethod` would return at that time.

**Comment by romani** [comment]
moment of evaluation (lazy evaluation) is a problem, and I do not see now good model where we can reliably be sure that update of method reference will be beneficial without any side effects.
I would rather make first version of Check without such property and focus on simple cases first , even there might be nuances.

**Comment by Vampire** [comment]
Well, as soon as you call any method or reference on any field, be it static or instance, you never know wether semantic changes, at least not without very extensive data flow analysis and even then you might not be able to tell.

Actually even if you only reference local variables, if those are arrays and you dereference the array, or if it is some other wrapper like `AtomicReference` and you dereference it, you have the same problem.

I would also vote to just ignore such lambdas for this check.
As soon as on the left side of the resulting method reference there is any method call, array index operator or field access, it should not be considered, as it could change semantic eventually.

If on the left side of the resulting method reference there are only local variables directly or classes, it should generate a violation.

**Comment by romani** [comment]
@strkkk , please find time to finish this PR.
I try to keep list of PR in active list only, I try to focus on PRs with high priority to reply quickly
If you see too much false positives it is better to declare them and close PR and issue.

If want to postpone this PR for a while, let's close this PR, you can reopen it at any time in future.

**Comment by strkkk** [comment]
Lets postpone it. I am ok to close it, I will reopen/create new PR later.

**Pull request closed without merging**